### PR TITLE
Anne Watson && Maria Velez - Octos - Scrabble

### PR DESCRIPTION
**Pull request by velezdev**
# Scrabble
Congratulations! You're submitting your assignment.

## Comprehension Questions

Feature | Feedback
--- | ---
What are the advantages of using git when collaboratively working on one code base? | We can merge our changes that are independent and see difference between versions. Easily resolve conflicts. Saves time. |
Why do you think we chose to make the `score` method in the `Scoring` class a **class method** instead of an **instance method**? | The 'score' method helps us see all instances of that class. Also we were able to use that class method in other classes. |
Describe an example of a test you wrote for a _nominal case_. | Most of them were for nominal cases. The 'adds input to the plays array' test added a word to the array of words at play for an instance of player.  | 
Describe an example of a test you wrote for an _edge case_. | One 'edge case' test was 'does not draw if the number passed is 0'; we would not expect for someone to request zero tiles. |
Does your code use any methods from the `Enumerable` mixin? If so, where and why was it helpful? | We used ".inject" in tiles_remaining method in TileBag. It was helpful because it let us add up all values in our hash of remaining tiles. |
What was one method you and your pair used to debug code? | We used lots of puts statements, pry, and break. |
Is there a specific piece of code you'd like feedback on? | All of it, if possible! Particularly if you have suggestions for edge case testing. |
What are two discussion points that you and your pair discussed when giving/receiving feedback from each other that you would be willing to share? | That we liked working together, that even though we were processing at different speeds the other was willing to stop and regroup to let the other know where we were at.|

### CODE REVIEW

File: lib/scoring.rb [hunk 1]
```diff
@@ -1,9 +1,89 @@
...
+        return nil
+      end
+
+      # Take input as a string, split the word, downcase it, store word in an array called "letter_array"
+      letter_array = []
+      letter_array = word.downcase.split(//)
+      # SCORECARD AKA 'provided_words_values'
+      value_1_array = ["a", "e", "i", "o", "u", "l", "n", "r", "s", "t"]
+      value_2_array = ["d", "g"]
+      value_3_array = ["b", "c", "m", "p"]
```
Comment (droberts-sea) [inline lib/scoring.rb:19]: I don't know that I like the idea of putting these all in separate arrays stored as different variables. Any time you have a bunch of different variables that are all working together to solve the same problem, you should think "How can I combine these together".

In this case, you could organize things as an array of arrays, or a hash mapping the point value to an array of letters. However, I think we could further re-organize things into a solution that's even easier to work with. If you store this data as a hash mapping a letter to its score, like:

```ruby
LETTERS = {
  "A" => 1
  "B" => 3,
  "C" => 3,
  "D" => 2,
  # ...
}
```

not only is it easy to read and modify, but you can get the score for a given letter by saying `LETTERS[letter]`, making the below code much simpler.

File: lib/scoring.rb [hunk 2]
```diff
@@ -1,9 +1,89 @@
...
+      # Highest scoring words
+      highest_scoring_words = scored_words_hash.select{|word,score| score == highest_score_value}
+      puts "This is the highest scoring word: #{highest_scoring_words}"
+
+      # Handles tie-breakers
+      if highest_scoring_words.length > 1
+        winner_hash = highest_scoring_words.select{|word| word.length == 7}
+        if winner_hash.empty?
+          shortest_winner_hash =  Hash[highest_scoring_words.sort_by {|word, score| word.length}]
+          return shortest_winner_hash.keys[0]
```
Comment (droberts-sea) [inline lib/scoring.rb:81]: Any time you sort a list and then take the first value, this can be simplified by using `max` or `min`. In this case, lines 80 and 81 could be shortened to `return highest_scoring_words.min_by { |word, score| word.length}.first`.

File: lib/scoring.rb [hunk 3]
```diff
@@ -1,9 +1,89 @@
...
+
+      array_of_words.each do |provided_word|
+        score = Scrabble::Scoring.score(provided_word)
+        scored_words_hash[provided_word] = score
+      end
+
+      # Highest score values
+      highest_score_array = scored_words_hash.max_by{|provided_word,score| score}
+      puts "This is the highest score value: #{highest_score_array[1]}"
+      highest_score_value = highest_score_array[1]
```
Comment (droberts-sea) [inline lib/scoring.rb:70]: This method works, but I think I would argue you're over-using enumerables here, and it makes the code more complex than it needs to be. My strategy, especially when I see that the enumerable methods aren't quite matching up to the problem I'm solving, is to write (or pseudocode) it as simply as I can using an `each` loop, and then try and spot patterns where an enumerable might make things cleaner.

Refer to the in-class implementation we did as an example of this.

File: specs/scoring_spec.rb [hunk 4]
```diff
@@ -42,21 +43,33 @@
...
     end
 
     it 'returns the only word in a length-1 array' do
+      one_word_array = ["apple"]
+      Scrabble::Scoring.highest_score_from(one_word_array).must_equal("apple")
     end
 
     it 'returns the highest word if there are two words' do
+      two_word_array = ["apple", "cat"] #apple = 9; cat = 5
+      Scrabble::Scoring.highest_score_from(two_word_array).must_equal("apple")
```
Comment (droberts-sea) [inline specs/scoring_spec.rb:57]: Like we discussed in class, it would probably be wise to check multiple orderings of these words, that is, both `['apple', 'cat']` and `['cat', 'apple']`. You'd be surprised how often this sort of thing turns up a bug.

File: specs/scoring_spec.rb [hunk 5]
```diff
@@ -42,21 +43,33 @@
...
     end
 
     it 'if tied, prefer a word with 7 letters' do
+      tie_array = ["lalalal", "foul"] #both 7 value
+      Scrabble::Scoring.highest_score_from(tie_array).must_equal("lalalal")
```
Comment (droberts-sea) [inline specs/scoring_spec.rb:62]: Since `lalalal` gets the 7-letter bonus, I don't think you actually have a tie here. There isn't a pair of English words that allow you to test this logic, but `zzzzzz` and `oratory` will do.

"Why bother" you ask? Well, if Miriam-Webster adds `zzzzzz` as an onomatopoeia for snoring, we need to be prepared, or you could envision a scrabble app that works across languages or that allows users to add their own words. Moreover, maybe this is the mathematician in me speaking, but I believe there's beauty and joy to be found in writing code that isn't just "good enough", but which truly solves the problem.

File: lib/player.rb [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+      return scores.sum
+    end
+
+    def won?
+      return total_score > 100
+    end
+
+    def highest_scoring_word
+      Scoring.highest_score_from
+    end
```
Comment (droberts-sea) [inline lib/player.rb:36]: Doesn't `highest_score_from` take an argument? Looks like you don't have any tests for this method, otherwise they would definitely have gotten an `ArgumentError`.

File: specs/player_spec.rb [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+      ada.play("cat").must_equal false
+    end
+
+    it "updates the total score after playing word" do
+      ada = Scrabble::Player.new("Ada")
+      ada.play("beets")
+      puts ada.total_score
+    end
+  end #ends describe 'play(word)'
+
```
Comment (droberts-sea) [inline specs/player_spec.rb:67]: You're missing tests for `Player#won?`. Here are some questions I would like to see answered by these tests:

- What if the player has not played any words at all?
- What if the player has less than 100 points?
- What if the player has exactly 100 points?
- What if the player has more than 100 points?

File: specs/player_spec.rb [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+  describe 'tiles' do
+    it 'holds 7 tiles the first time a player draws tiles' do
+      ada = Scrabble::Player.new("Ada")
+      game_tiles = Scrabble::TileBag.new
+      ada.draw_tiles(game_tiles)
+      ada.tiles.length.must_equal(7)
+    end
```
Comment (droberts-sea) [inline specs/player_spec.rb:74]: Maybe a test about if a word is played, the tiles for that word go away?

File: lib/tilebag.rb [hunk 9]
```diff
@@ -0,0 +1,51 @@
...
+        x: 1,
+        y: 2,
+        z: 1
+      }
+    end
+
+    def draw_tiles(num)
+      drawn_tiles = @tile_bag.to_a.sample(num)
+      drawn_tiles.each do |letter_frequency|
+        @tile_bag[letter_frequency[0]] -=1
```
Comment (droberts-sea) [inline lib/tilebag.rb:40]: By this logic, you're as likely to draw an `e` as you are a `z`! It also means that you won't ever get more than one of the same letter, which could be bad if you need to draw the last two tiles from the bag and they're both `i`. However, the way you've set up your data structures, it would be difficult to do it any other way.

Instead, you might keep all the letters in the bag in an array. You could seed the array with code like:

```ruby
def initialize
  letters = {
    # the hash you had before
  }
  @tiles = []
  letters.each do |letter, count|
    count.times do
      @tiles << letter
    end
  end
end
```

Then you could pull a random tile with `sample` and get a nice even distribution.

File: specs/tilebag_spec.rb [hunk 10]
```diff
@@ -0,0 +1,56 @@
...
+    end
+
+    it 'decrements the amount of letters remaining for a letter tile that is drawn' do
+      game_1_tiles = Scrabble::TileBag.new
+      before = game_1_tiles.tile_bag.clone
+      drawn_tiles = game_1_tiles.draw_tiles(1)
+      symbol = drawn_tiles[0][0]
+      (before[symbol] - game_1_tiles.tile_bag[symbol]).must_equal 1
+    end
+  end
```
Comment (droberts-sea) [inline specs/tilebag_spec.rb:46]: Other questions I'd like to see answered by these tests:

- Am I able to draw the last tile from the bag?
- What happens if I try to draw more tiles than are currently in the bag?

**Comment by droberts-sea** [comment]
# Scrabble
## What We're Looking For

Feature | Feedback
--- | ---
**General** |
Answered comprehension questions | yes
Both Teammates contributed to the codebase | yes
Regular Commits with meaningful commit messages | yes - good work!
**Scoring class** |
`score` method | yes
Uses appropriate data structure to store the letter score | see inline
Appropriately handles edge cases | yes
`highest_score_from` method | yes
Appropriately handles edge cases | yes
Test implementations for `highest_score_from`| mostly - see inline
**Player class** |
Uses an array to keep track of words played | yes
Uses existing code to create `highest_scoring_word` and `highest_word_score` | kind of - you're missing `highest_word_score`, and see inline comment on `highest_scoring_word`
Returns `true` or `false` from `won?` method | yes
Tests edge cases on the `play(word)` | some missing - see inline
**TileBag class** |
Uses a data structure to keep the set of default tiles in the bag | see inline
`draw_tiles` method uses the parameter to draw a set of random tiles | yes
`tiles_remaining` method returns a count | yes
Tests `draw_tiles` edge cases | see inline

Good job overall! The big learning goals I'm looking to see demonstrated on this project are:

- Building complex logic (tie-breaking logic in `highest_score_from`)
- Test coverage
- Object composition (interaction between `Player` and `TileBag`)

I feel that you've done a good job of demonstrating the first and the third of these. Your test coverage could use some work - several of your methods are missing tests entirely, or are missing interesting cases. I've left some inline comments for you to review, please do so, and keep up the hard work!
